servicetalk-loadbalancer-experimental/src/main/java/io/servicetalk/loadbalancer/ConnectionPoolPolicies.java

```diff
@@ -17,6 +17,9 @@
 
 import io.servicetalk.client.api.LoadBalancedConnection;
 
+/**
+ * A factory to create different {@link ConnectionPoolPolicy} variants.
+ */
 public final class ConnectionPoolPolicies {
```

Comment:
I think what you're proposing is that if we keep the abstract class and introduce an interface later we can use the good name (`ConnectionSelectionPolicy`) for the interface. Eg, not trying to convert the abstract class to an interface which I'm not sure is considered API compatible. Maybe it is since there is no way for users to create one, but otoh I'm not sure.

Reply:
You are right, if we don't allow users create their own implementations there is almost 0 risk to break them when we change from an abstract class to an interface. My only worry is that if someone releases a library that references `ConnectionPoolPolicy` and assigns it to a variable, will that library work without recompilation with the next version of ST if we change it to an interface? I don't remember binary compatibility rules in this case.

Reply:
I ran the idea of converting the abstract class to an interface by japicmp and unfortunately it gives a 👎 
```
***! MODIFIED INTERFACE (<- CLASS) : PUBLIC ABSTRACT io.servicetalk.loadbalancer.LoadBalancingPolicy  (not serializable)
	===  CLASS FILE FORMAT VERSION: 52.0 <- 52.0
	GENERIC TEMPLATES: === C:io.servicetalk.client.api.LoadBalancedConnection, === ResolvedAddress:java.lang.Object
	===  UNCHANGED SUPERCLASS: java.lang.Object (<- java.lang.Object)
```

Reply:
How about this: right now the `abstract class` version is essentially just a marker interface as far as users are concerned as the `buildSelector(..)` is the private piece and contains the types that we want to keep private for now. We could make the policy types into marker interfaces for the public API and simply extend it for the internal details. When we go to make calls to `buildSelector(..)`, which is currently the only functional aspect we use, we can downcast. This means users can technically extend the interface but it will be of no practical use and doesn't leak the types we're not ready to commit to yet.

Reply:
Here is a WIP of what that would look like for one of these types: https://github.com/apple/servicetalk/pull/3144

Reply:
We tested that users bytecode doesn't change if they only reference the type:
```java
        ConnectionPoolPolicy<LoadBalancedConnection> policy = ConnectionPoolPolicies.linearSearch();
        LoadBalancerBuilder<Object, LoadBalancedConnection> builder = LoadBalancers.builder("test")
                .connectionPoolPolicy(policy);
```
We can change it to an interface later and it will be binary compatible change for users. So we decided to rename it now to `ConnectionSelectorPolicy`: https://github.com/apple/servicetalk/pull/3145